### PR TITLE
Feat: [OV-43017] Rich Text Editor

### PR DESCRIPTION
**Pull request by vicky-comeau**

### CODE REVIEW

File: packages/RichTextEditor/src/RichTextEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+  /** The initial editor state, which will preload the editor with content */
+  initialState?: string;
+  /** Whether or not the text can be cleared */
+  isClearable?: boolean;
+  /** Whether or not the rich text editor should have a private look */
+  isPrivate?: boolean;
+  /** The maximum amount or characters allowed
+   * (Automatically adds a character limit indicator) */
+  maxLength?: number;
+  /** Messages for the buttons */
```
Comment (franckgaudin) [inline packages/RichTextEditor/src/RichTextEditor.tsx:86]: You should be a little more explicit on the description, at first I thought it was the text for the save button 🙃

File: packages/RichTextEditor/package.json [hunk 2]
```diff
@@ -0,0 +1,50 @@
+{
+  "name": "@igloo-ui/rich-text-editor",
```
[the remark this answers is not in the document]
Comment (vicky-comeau) [reply packages/RichTextEditor/package.json:2]: I agree, I hated writing this everywhere haha. I can change it to text-editor.

File: packages/RichTextEditor/src/rich-text-editor.scss [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+  --ids-rich-text-editor-bold: #{tokens.$font-weight-bold};
+  
+  /* Text Code */
+  --ids-rich-text-editor-text-code-background: #{tokens.$grey-200};
+  
+  /* Link */
+  --ids-rich-text-editor-link-color: #{tokens.$electric-blue-500};
+  --ids-rich-text-editor-link-font-weight: #{tokens.$font-weight-medium};
+  
+  /* Code */
```
[the remark this answers is not in the document]
Comment (vicky-comeau) [reply packages/RichTextEditor/src/rich-text-editor.scss:37]: For the display of link or code? Well the Code block only shows up if you know the shortcut. I just followed what Lexical had in their example and tried to match it to the colors we had. There is nothing int he design for this.

File: packages/RichTextEditor/src/rich-text-editor.scss [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+}
+
+.ids-rich-text-editor__inner {
+  line-height: var(--ids-rich-text-editor-line-height);
+  position: relative;
+}
+
+.ids-rich-text-editor__scroller {
+  display: flex;
+  min-height: 50px;
```
[the remark this answers is not in the document]
Comment (vicky-comeau) [reply packages/RichTextEditor/src/rich-text-editor.scss:121]: I'll switch it

File: packages/RichTextEditor/src/rich-text-editor.scss
[the remark this answers is not in the document]
Comment (vicky-comeau) [reply packages/RichTextEditor/src/rich-text-editor.scss:1]: I will set .ids-text-editor to 12rem, which seems to be the height in figma. So if the toolbar and and footer area is not displayed, there will be more editable content displayed and it'll still equal the height of an editor that has a toolbar and footer.

File: packages/TextEditor/src/TextEditor.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,83 @@
+import React from 'react';
+
+import { Meta, StoryFn, StoryObj } from '@storybook/react';
```
Comment (ofrogon) [inline packages/TextEditor/src/TextEditor.stories.tsx:3]: `StoryFn` is unused

```suggestion
import { Meta, StoryObj } from '@storybook/react';
```

File: packages/TextEditor/src/TextEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+  maxLength?: number;
+  /** Messages for buttons, tooltips and text to localize the text editor */
+  messages?: Messages;
+  /** The event that is called when the editor loses focus */
+  onBlur?: (editor?: LexicalEditor) => void;
+  /** The event that is called when the editor text changes */
+  onChange?: (editorState: string) => void;
+  /** The event that's called when the editor gains focus */
+  onFocus?: (editor?: LexicalEditor) => void;
+  /** The placeholder text to display int he editor */
```
Comment (ofrogon) [inline packages/TextEditor/src/TextEditor.tsx:94]:
```suggestion
  /** The placeholder text to display in the editor */
```

File: packages/TextEditor/src/TextEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+  placeholder?: string;
+  /** Usually the save button for the editor */
+  primaryBtn?: React.ReactElement;
+  /** Wether or not the editor should display the toolbar */
```
Comment (ofrogon) [inline packages/TextEditor/src/TextEditor.tsx:98]:
```suggestion
  /** Whether or not the editor should display the toolbar */
```

File: packages/TextEditor/src/TextEditor.tsx [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+  showToolbar?: boolean;
+  /** Wether or not the editor should display
```
Comment (ofrogon) [inline packages/TextEditor/src/TextEditor.tsx:100]:
```suggestion
  /** Whether or not the editor should display
```

File: packages/TextEditor/src/plugins/MaxLengthPlugin.tsx [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+export function MaxLengthPlugin({ maxLength }: { maxLength: number }): null {
+  const [editor] = useLexicalComposerContext();
+
+  useEffect(() => {
+    let lastRestoredEditorState: EditorState | null = null;
+
+    return editor.registerNodeTransform(RootNode, (rootNode: RootNode) => {
+      const selection = $getSelection();
+      if (!$isRangeSelection(selection) || !selection.isCollapsed()) {
```
Comment (ofrogon) [inline packages/TextEditor/src/plugins/MaxLengthPlugin.tsx:21]: Is collapse is a property, not a method given to lib.dom.d.ts (typescript)
image

File: packages/TextEditor/src/plugins/ToolbarPlugin.tsx [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+  messages?: Messages;
+  /** Whether or not the toolbar should show on focus and hide on blur */
+  showOnFocus?: boolean;
+}
+
+function Divider(): JSX.Element {
+  return <div className="ids-toolbar__divider" />;
+}
+
+export interface FloatingLinkEditorProps {
```
Comment (ofrogon) [inline packages/TextEditor/src/plugins/ToolbarPlugin.tsx:63]: `FloatingLinkEditorProps` is unused

File: packages/TextEditor/src/plugins/ToolbarPlugin.tsx [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+import {
+  SELECTION_CHANGE_COMMAND,
+  FORMAT_TEXT_COMMAND,
+  CLEAR_EDITOR_COMMAND,
+  $getSelection,
+  $isRangeSelection,
+  $getRoot,
+  $isParagraphNode,
+  COMMAND_PRIORITY_CRITICAL,
+  LexicalEditor,
```
Comment (ofrogon) [inline packages/TextEditor/src/plugins/ToolbarPlugin.tsx:18]: Can also be remove with `FloatingLinkEditorProps`

File: packages/TextEditor/src/text-editor.scss [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+  --ids-text-editor-code-before-color: #{tokens.$grey-600};
+  --ids-text-editor-code-before-padding: #{tokens.$space-2};
+  --ids-text-editor-code-after-font-size: #{tokens.$font-size-1};
+  
+  /* Code Tokens */
+  --ids-text-editor-token-comment: #{tokens.$grey-600};
+  --ids-text-editor-token-punctuation: #{tokens.$grey-500};
+  --ids-text-editor-token-property: #{tokens.$coral-900};
+  --ids-text-editor-token-selector: #{tokens.$seaweed-800};
+  --ids-text-editor-token-variable: #{tokens.$dandelion-600};
```
Comment (ofrogon) [inline packages/TextEditor/src/text-editor.scss:56]: `--ids-text-editor-token-variable` is unused

File: packages/TextEditor/src/text-editor.scss [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+  padding: 0;
+  margin: var(--ids-text-editor-quote-spacing);
+}
+  
+.ids-text-editor__list-ul {
+  padding: 0;
+  margin: var(--ids-text-editor-quote-spacing);
+}
+  
+.ids-text-editor__listitem {
```
Comment (ofrogon) [inline packages/TextEditor/src/text-editor.scss:361]: `list-item` or `listItem`?

File: packages/TextEditor/src/text-editor.scss [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+  margin: var(--ids-text-editor-list-margin);
+}
+  
+.ids-text-editor__nested-listitem {
```
Comment (ofrogon) [inline packages/TextEditor/src/text-editor.scss:365]: `list-item` or `listItem`?

File: packages/TextEditor/src/utils/url.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
+export function sanitizeUrl(url: string): string {
+  /** A pattern that matches safe  URLs. */
```
Comment (ofrogon) [inline packages/TextEditor/src/utils/url.ts:2]:
```suggestion
  /** A pattern that matches safe URLs. */
```

File: utils/bootstrap/helpers/template.mjs [hunk 16]
```diff
@@ -59,7 +59,7 @@ export const getStoriesTemplate = (name) => {
 
   return `import React from 'react';
 
-  import { ComponentMeta, ComponentStory } from '@storybook/react';
+  import { Meta, StoryFn, StoryObj } from '@storybook/react';
```
Comment (ofrogon) [inline utils/bootstrap/helpers/template.mjs:62]: `StoryFn` is unused here

File: packages/TextEditor/src/plugins/MaxLengthPlugin.tsx (hunk 9 above)
Comment (vicky-comeau) [reply packages/TextEditor/src/plugins/MaxLengthPlugin.tsx:21]: It's defined in LexicalSelection as a method: 
image
Also this was taken from lexical's playground plugins: https://github.com/facebook/lexical/blob/main/packages/lexical-playground/src/plugins/MaxLengthPlugin/index.tsx#L23

File: packages/TextEditor/src/text-editor.scss (hunk 13 above)
Comment (vicky-comeau) [reply packages/TextEditor/src/text-editor.scss:361]: I'll go with list-item and I'll change the others to match.

**Review by franckgaudin — APPROVED** [review]

**Pull request merged**